src/array_api_extra/_apply.py

```diff
@@ -0,0 +1,351 @@
+        (1,)]``.
+        Default: disallow Dask.
+    core_indices : Sequence[Sequence[Hashable]], optional
+        **Dask specific.**
+        Axes labels of each input array that cannot be broken into chunks.
+        Default: disallow Dask.
+    output_indices : Sequence[Sequence[Hashable]], optional
+        **Dask specific.**
+        Axes labels for each output array. If `func` returns a single (non-sequence)
+        output, this must be a sequence containing a single sequence of labels, e.g.
+        ``['ijk']``.
+        Default: disallow Dask.
+    adjust_chunks : Sequence[Mapping[Hashable, Callable[[int], int]]], optional
+        **Dask specific.**
+        Sequence of dicts, one per output, mapping index to function to be applied to
+        each chunk to determine the output size. The total must add up to the output
+        shape.
+        Default: on Dask, the size along each index cannot change.
+    new_axes : Sequence[Mapping[Hashable, int]], optional
+        **Dask specific.**
+        New indexes and their dimension lengths, one per output.
+        Default: on Dask, there can't be `output_indices` that don't appear in
+        `input_indices`.
+    **kwargs : Any, optional
+        Additional keyword arguments to pass verbatim to `func`.
+        Any array objects in them won't be converted to NumPy.
+
+    Returns
+    -------
+    tuple[Array, ...]
+        The result(s) of `func` applied to the input arrays.
+        This is always a tuple, even if `func` returns a single output.
+
+    Notes
+    -----
+    JAX:
+        This allows applying eager functions to jitted JAX arrays, which are lazy.
+        The function won't be applied until the JAX array is materialized.
+
+        The `JAX transfer guard
+        <https://jax.readthedocs.io/en/latest/transfer_guard.html>`_
+        may prevent arrays on a GPU device from being transferred back to CPU.
+        This is treated as an implicit transfer.
+
+    PyTorch, CuPy:
+        These backends raise by default if you attempt to convert arrays on a GPU device
+        to NumPy.
+
+    Dask:
+        This function allows applying func to the chunks of dask arrays.
+        The dask graph won't be computed. As a special limitation, `func` must return
+        exactly one output.
+
+        In order to allow Dask you need to specify at least
+        `input_indices`, `output_indices`, and `core_indices`, but you may also need
+        `adjust_chunks` and `new_axes` depending on the function.
+
+        Read `dask.array.blockwise`:
+        - ``input_indices`` map to the even ``*args`` of `dask.array.blockwise`
+        - ``output_indices[0]`` maps to the ``out_ind`` parameter
+        - ``adjust_chunks[0]`` maps to the ``adjust_chunks`` parameter
+        - ``new_axes[0]`` maps to the ``new_axes`` parameter
+
+        ``core_indices`` is a safety measure to prevent incorrect results on
+        Dask along chunked axes. Consider this::
```

Comment:
This design was informed from https://docs.xarray.dev/en/latest/generated/xarray.apply_ufunc.html